File: competitions/weekofcode28/white.cpp
#include <iostream>
#include <unordered_map>
#include <cmath>
#include <utility>
#include <iomanip>

using namespace std;

struct SHash {
    size_t operator()(const pair<int, int>& p) const {
        return p.first ^ p.second;
    }
};

unordered_map< pair<int,int>, double, SHash > dp;

int reverse(int b, int n) {
    int sum = 0;
    for (int i = 0; i < n; i++) {
        if ((b >> i) & 1) {
            sum += (1 << (n-1-i));
        }
    }
    return sum;
}

int get_bit(int b, int i) {
    return (b >> i) & 1;
}

int remove_bit(int x, int n, int i) {
    int mask = (int)pow(2, (i+1)) - 1;
    int bits = x & ~mask;
    x &= ~(1<<i);
    x &= ~(1<<n-1);
    return (bits >> 1) | (x & mask);
}

double white(int b, int n, int k) {
    pair<int,int> args(b, k);
    unordered_map< pair<int,int>, double, SHash>::iterator it = dp.find(args);
    if (it != dp.end()) {
        return (*it).second;
    }

    double ev;
    if (k == 1) { 
        int summ = 0;
        int lo = 0; int hi = n - 1;
        while (lo <= hi) {
            if (get_bit(b, lo) || get_bit(b, hi)) {
                summ += lo == hi ? 1 : 2;
            } 
            lo++; hi--;
        }
        ev = (double)summ/n;
    } else {
        ev = 0.0;
        int lo = 0; int hi = n - 1;
        while (lo <= hi) {
            double lo_ev = white(remove_bit(b, n, lo), n-1, k-1);
            if (get_bit(b, lo) == 1) lo_ev += 1;
            if (lo == hi) {
                ev += lo_ev;
            } else {
                double hi_ev = white(remove_bit(b, n, hi), n-1, k-1);
                if (get_bit(b, hi) == 1) hi_ev += 1;
                ev += 2*(lo_ev > hi_ev ? lo_ev : hi_ev);
            }
            lo++; hi--;
        }
        ev = ev/(double)n;
    }

    dp[args] = ev;
    dp[make_pair(reverse(b, n), k)] = ev;
    return ev;
}

int main() {
    int n,k;
    cin >> n >> k;
    char c;
    int b;
    cin.get(c);
    for (int i = 0; i < n; i++) {
        cin.get(c);
        if (c == 'W') {
            b |= (int)pow(2, i);
        }
    }
    
    double ans = white(b, n, k);
    cout << fixed << showpoint << setprecision(10) << ans << endl;
}

File: competitions/weekofcode28/white.py
#!/bin/python3

def white(b, n, k, sumb, dp):
    try: return dp[(b, k)]
    except KeyError: pass

    if sumb == n:
        ev = float(k)
    elif sumb == 0:
        ev = 0.0
    elif k == 1:
        summ = 0
        lo = 0; hi = n - 1
        while lo <= hi:
            v = max(b[lo], b[hi])
            if lo != hi: v *= 2
            summ += v
            lo += 1; hi -= 1
        ev = summ/n
    else:
        ev = 0
        lo = 0; hi = n - 1
        while lo <= hi:
            lo_ev = white(b[:lo] + b[lo+1:], n-1, k-1, sumb - b[lo], dp)
            if b[lo] == 1: lo_ev += 1
            if lo == hi:
                ev += lo_ev
            else:
                hi_ev = white(b[:hi] + b[hi+1:], n-1, k-1, sumb - b[hi], dp)
                if b[hi] == 1: hi_ev += 1
                ev += 2*max(lo_ev, hi_ev)
            lo += 1; hi -= 1
        ev /= n

    dp[(b, k)] = dp[(tuple(reversed(b)), k)] = ev
    return ev

def main():
    n, k = map(int, input().split())
    b = tuple([1 if ball == 'W' else 0 for ball in input()])

    dp = dict()

    print(white(b, n, k, sum(b), dp))

main()
